plan/expression_rewriter.go

```diff
@@ -740,6 +743,23 @@ func datumToConstant(d types.Datum, tp byte) *expression.Constant {
 	return &expression.Constant{Value: d, RetType: types.NewFieldType(tp)}
 }
 
+// paramToExpression rewrites ParamMarkerExpr to a function expression, GetParam.
+// GetParam function should be evaluated whenever execute the statement because
+// the parameter values are given just when executing the prepared statement generally.
+func (er *expressionRewriter) paramToExpression(v *ast.ParamMarkerExpr) {
```

Comment:
since you replace the place holder "?" to a function call "GetParam", which will return a constant, maybe you should do the type inference of expressions to add cast functions to avoid type mismatch on the plan got from the prepared plan cache ?

Reply:
Since the type of parameter values is always VarString, the return type of GetVar is also VarString. Thus, additional type conversions are not required when using the plan cached. 

However, if typed parameters will be supported, the additional type conversions will be added as you mentioned.